### PR TITLE
[BEAM-559] Support for @Setup and @Teardown in DoFnTester

### PR DESCRIPTION
**Pull request by jkff**
- To allow testing DoFn reuse, adds a third cloning mode - CLONE_ONCE. The current CLONE is renamed to CLONE_PER_BUNDLE because it corresponds to the olden ways of DoFnTester. This rename should be harmless because it was the default, so probably nobody uses it (nobody uses it in the Beam codebase anyway), but I can rename it back if necessary.
- Calls Setup lazily on the first interaction with the fn under test: depending on cloning mode, this is either a per-bundle clone, or the original fn, or a single clone of the original fn.
- Calls Teardown either when the DoFnTester is closed, or when finishBundle'ing with CLONE_PER_BUNDLE.
- Enforces that you have to configure the DoFnTester (setSideInputs, setCloningBehavior) before using it, to simplify the semantics. Makes setSideInputs consistent with setSideInput(without "s") in not resetting the state.
- Allows processing a bundle after processing an element. This used to be disallowed, but it's weird, because processing a bundle after processing a bundle *was* allowed.

Additionally, improves testing coverage.

### CODE REVIEW

**Comment by jkff** [comment]
R: @tgroh

**Comment by dhalperi** [comment]
(driveby: I assigned BEAM-559 to you; can you please add this to PR message?)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 1]
```diff
@@ -78,10 +78,10 @@
  * @param <InputT> the type of the {@link DoFn}'s (main) input elements
  * @param <OutputT> the type of the {@link DoFn}'s (main) output elements
  */
-public class DoFnTester<InputT, OutputT> {
+public class DoFnTester<InputT, OutputT> implements AutoCloseable {
   /**
    * Returns a {@code DoFnTester} supporting unit-testing of the given
-   * {@link DoFn}.
+   * {@link DoFn}. By default, uses {@link CloningBehavior#CLONE_PER_BUNDLE}.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Should be `CLONE_ONCE` to more accurately represent most backend behaviors.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 2]
```diff
@@ -108,8 +110,8 @@
    * {@link DoFn} takes no side inputs.
    */
   public void setSideInputs(Map<PCollectionView<?>, Map<BoundedWindow, ?>> sideInputs) {
+    checkState(state == State.UNINITIALIZED, "Wrong state: %s", state);
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: "Can't add side inputs after `DoFnTester` is initialized: in state %s"?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/DoFnTesterTest.java [hunk 3]
```diff
@@ -51,122 +53,180 @@
...
-    take = tester.takeOutputElements();
-    assertThat(take, hasItems("5", "6"));
-
-    tester.finishBundle();
-
-    // verify finishBundle() was called.
-    assertTrue(deserializedDoFn.wasStartBundleCalled());
-    assertTrue(deserializedDoFn.wasFinishBundleCalled());
+    for (DoFnTester.CloningBehavior cloning : DoFnTester.CloningBehavior.values()) {
+      try (DoFnTester<Long, String> tester = DoFnTester.of(new CounterDoFn())) {
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/DoFnTesterTest.java]: This is a really good idiom.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java (hunk 1 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: I used CLONE_PER_BUNDLE to more accurately represent the previous behavior of DoFnTester (to minimize breakages in people's pre-existing tests) but I guess this makes more sense, and I have a hard time imagining what kind of code one would write that would actually break.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java (hunk 2 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 4]
```diff
@@ -695,13 +719,26 @@ public void outputWithTimestamp(OutT output, Instant timestamp) {
...
   /////////////////////////////////////////////////////////////////////////////
 
   /** The possible states of processing a {@link DoFn}. */
-  enum State {
-    UNSTARTED,
-    STARTED,
-    FINISHED
+  private enum State {
+    UNINITIALIZED,
+    INSIDE_BUNDLE,
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Done.

**Comment by dhalperi** [comment]
Merged based on Thomas' LGTM.

**Comment by dhalperi** [comment]
(and I looked at the travis flake, and it's still a flake).

**Pull request closed without merging**